libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,995 @@
+	auto const [endLine, endColumn] = _location.source->translatePositionToLineColumn(_location.end);
+	return toJsonRange(startLine, startColumn, endLine, endColumn);
+}
+
+Json::Value toJson(boost::filesystem::path const& _basePath, SourceLocation const& _location)
+{
+	solAssert(_location.source.get() != nullptr, "");
+	Json::Value item = Json::objectValue;
+	item["uri"] = toFileURI(_basePath / boost::filesystem::path(_location.source->name()));
+	item["range"] = toJsonRange(_location);
+	return item;
+}
+
+std::pair<size_t, size_t> offsetsOf(std::string const& _text, LineColumnRange _range) noexcept
+{
+	auto const start = CharStream::translateLineColumnToPosition(_text, _range.start.line, _range.start.column);
+	auto const end = CharStream::translateLineColumnToPosition(_text, _range.end.line, _range.end.column);
+	solAssert(start.has_value(), "");
+	solAssert(end.has_value(), "");
+	return std::pair{ size_t(start.value()), size_t(end.value()) };
+}
+
+} // }}} end helpers
+
+LanguageServer::LanguageServer(Logger _logger, unique_ptr<Transport> _transport):
+	m_client{move(_transport)},
+	m_handlers{
+		{"$/cancelRequest", {} }, // Don't do anything for now, as we're synchronous.
+		{"cancelRequest", {} }, // Don't do anything for now, as we're synchronous.
+		{"initialize", bind(&LanguageServer::handle_initialize, this, _1, _2)},
+		{"initialized", {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"textDocument/definition", bind(&LanguageServer::handle_textDocument_definition, this, _1, _2)},
+		{"textDocument/didChange", bind(&LanguageServer::handle_textDocument_didChange, this, _1, _2)},
+		{"textDocument/didClose", [](auto, auto) {/*nothing for now*/}},
+		{"textDocument/didOpen", bind(&LanguageServer::handle_textDocument_didOpen, this, _1, _2)},
+		{"textDocument/documentHighlight", bind(&LanguageServer::handle_textDocument_highlight, this, _1, _2)},
+		{"textDocument/hover", bind(&LanguageServer::handle_textDocument_hover, this, _1, _2)},
+		{"textDocument/implementation", bind(&LanguageServer::handle_textDocument_implementation, this, _1, _2)},
+		{"textDocument/references", bind(&LanguageServer::handle_textDocument_references, this, _1, _2)},
+		{"workspace/didChangeConfiguration", bind(&LanguageServer::handle_workspace_didChangeConfiguration, this, _1, _2)},
+	},
+	m_logger{move(_logger)}
+{
+}
+
+DocumentPosition LanguageServer::extractDocumentPosition(Json::Value const& _json) const
+{
+	DocumentPosition dpos{};
+
+	dpos.path = extractPathFromFileURI(_json["textDocument"]["uri"].asString()).value();
+
+	if (boost::algorithm::starts_with(dpos.path, m_basePath.generic_string()))
+		dpos.path = dpos.path.substr(m_basePath.generic_string().size());
+
+	dpos.position.line = _json["position"]["line"].asInt();
+	dpos.position.column = _json["position"]["character"].asInt();
+
+	return dpos;
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	if (_settings["evm"].isString())
```

Comment:
Is this the same settings object that is used by standard-json? Why do we need to parse it here?

Reply:
just to answer that in here... no it's not. I am trying to reduce code dup though by looking at what I can reuse in the `StandardCompiler` interface. Not too much, and it need(ed) a little refactor to expose some sub logic. I think it was still worth doing so (WIP).